Source/WTF/wtf/ThreadSafeRefCounted.h

```diff
@@ -140,41 +140,39 @@ template<class T, DestructionThread destructionThread = DestructionThread::Any>
         if (!derefBase())
             return;
 
-        auto deleteThis = [this] {
+        if constexpr (destructionThread == DestructionThread::Any) {
             delete static_cast<const T*>(this);
-        };
-        switch (destructionThread) {
-        case DestructionThread::Any:
-            break;
-        case DestructionThread::Main:
-            ensureOnMainThread(WTFMove(deleteThis));
-            return;
-        case DestructionThread::MainRunLoop:
-            ensureOnMainRunLoop(WTFMove(deleteThis));
-            return;
+        } else if constexpr (destructionThread == DestructionThread::Main) {
+            ensureOnMainThread([this] {
+                delete static_cast<const T*>(this);
+            });
+        } else if constexpr (destructionThread == DestructionThread::MainRunLoop) {
+            ensureOnMainRunLoop([this] {
+                delete static_cast<const T*>(this);
+            });
+        } else {
+            static_assert(destructionThread == DestructionThread::Any || destructionThread == DestructionThread::Main || destructionThread == DestructionThread::MainRunLoop);
```

Comment:
I think your intention here is to prove that this branch is unreachable, by virtue of the fact that, if this assertion is true, then some branch above must have already been taken.

I think this formulation is a bit sketchy as-is because, though a static_assert is enforced at compile time, the relationship between what we're asserting and what we want to know is not. There is no proof that our assertion exhaustively lists each value for DestructionThread, and there is no proof that each value for DestructionThread is listed above as an independent branch condition.

Since there's no such thing as `constexpr switch` or `static_assert_unreachable`, I think the best you can do here is RELEASE_ASSERT_NOT_REACHED.

Reply:
I'd be hesitant to change a compile-time assertion into a runtime one, which may delay the realization that an error was introduced.
The purer `static_assert(false)` doesn't work on some compilers, as they just compile everything that doesn't depend on a template parameter, so it triggers even when the branch is not taken!
But I could change it into something like `static_assert(sizeof(destructionThread) == 0, "Unexpected destructionThread enumerator")`, it's a bit ugly (suggestions?) but it should catch any instantiation of this branch.
How about that?

Reply:
Relatedly, I've just found this c++ proposal: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2593r0.html "Allowing static_assert(false)" 😄 (I'm using the first "terrible" and "wrong" but "get the job done" workaround.)

Reply:
That didn't work, some compilers are too clever! I'll try other things...

Reply:
Ok, this holds.

> there's no such thing as [...] static_assert_unreachable

We could introduce something like `STATIC_ASSERT_NOT_REACHED_WITH_DEPENDENT_TYPE(T)` (with a better implementation than what I have here), would you like that? If yes, in this patch or separate?

Reply:
No news = good news? 😄
Since my current `static_assert` is pretty much equivalent to `RELEASE_ASSERT_NOT_REACHED` (but at compile time), I'll land this now. If I hear complaints we can always patch again.
And I'll prepare a separate patch for `STATIC_ASSERT_NOT_REACHED_WITH_DEPENDENT_TYPE`.